llama_index/program/predefined/df.py

```diff
@@ -0,0 +1,222 @@
+from llama_index.types import Model
+
+from typing import Type
+
+from typing import Optional, List, Any, Type
+from pydantic import BaseModel, Field
+from llama_index.types import BaseOutputParser
+from llama_index.program.base_program import BasePydanticProgram
+from llama_index.program.openai_program import OpenAIPydanticProgram
+from llama_index.prompts.prompts import Prompt
+from llama_index.program.llm_prompt_program import BaseLLMPromptProgram
+import pandas as pd
+
+
+class DataFrameRow(BaseModel):
+    """Row in a DataFrame."""
+
+    row_values: List[Any] = Field(
+        ...,
+        description="List of row values, where each value corresponds to a row key.",
+    )
+
+
+class DataFrameColumn(BaseModel):
+    """Column in a DataFrame."""
+
+    column_name: str = Field(..., description="Column name.")
+    column_desc: Optional[str] = Field(..., description="Column description.")
+
+
+class DataFrame(BaseModel):
+    """Data-frame class.
+
+    Consists of a `rows` field which is a list of dictionaries,
+    as well as a `columns` field which is a list of column names.
+
+    """
+
+    description: Optional[str] = None
+
+    columns: List[DataFrameColumn] = Field(..., description="List of column names.")
+    rows: List[DataFrameRow] = Field(
+        ...,
+        description="""List of DataFrameRow objects. Each DataFrameRow contains \
+        valuesin order of the data frame column.""",
+    )
+
+    def to_df(self) -> pd.DataFrame:
+        """To dataframe."""
+        return pd.DataFrame(
+            [row.row_values for row in self.rows],
+            columns=[col.column_name for col in self.columns],
+        )
+
+
+class DataFrameRowsOnly(BaseModel):
+    """Data-frame with rows. Assume column names are already known beforehand."""
+
+    rows: List[DataFrameRow] = Field(..., description="""List of row objects.""")
+
+    def to_df(self, existing_df: Optional[pd.DataFrame] = None) -> pd.DataFrame:
+        """To dataframe."""
+        if existing_df is None:
+            return pd.DataFrame([row.row_values for row in self.rows])
+        else:
+            new_df = pd.DataFrame([row.row_values for row in self.rows])
+            new_df.columns = existing_df.columns
+            # assume row values are in order of column names
+            return existing_df.append(new_df, ignore_index=True)
+
+
+DEFAULT_FULL_DF_PARSER_TMPL = """
+Please extract the following query into a structured data.
+Query: {input_str}.
+Please extract both the set of column names and row names.
+"""
+
+DEFAULT_ROWS_DF_PARSER_TMPL = """
+Please extract the following query into structured data.
+Query: {input_str}.
+The column schema is the following: {column_schema}.
+"""
+
+
+class DFFullProgram(BasePydanticProgram[DataFrame]):
+    """Data-frame program.
+
+    Extracts text into a schema + datapoints.
+
+    """
+
+    def __init__(
+        self,
+        pydantic_program_cls: Type[BaseLLMPromptProgram],
```

Comment:
So just to clarify -> `DFFullProgram` is a pydnatic program, which can use either a nested completion or function program under the hood to extract the DF stuff?

Reply:
Also was a bit confused here. Thought it just needs a `BasePydanticProgram` under the hood.

Reply:
so the issue with a BasePydanticProgram is that if we pass in just the class, we don't really have a way to initialize it since BasePydanticProgram is just a general "chain" interface for literally anything.

Initializing a BaseLLMPromptProgram can be a bit more defined since it's a type of Program that specifically takes in an output class, prompt, and llm. So we have a concrete interface to initialize this.

At a high-level, I was trying to figure out a good UX for DFProgram to not be tied to OpenAI or Guidance, but also have a good way of "initializing" a prokgram under the hood